Add FXIOS-12720 Update splash screen

razvanlitianu opened the pull request:
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-12720)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/27706)

## :bulb: Description
<!--- Describe your changes so the reviewer can understand the context and decisions made for your work -->

## :movie_camera: Demos
<!-- Please upload screenshots or video demos of your work, if applicable -->
<!-- You can either use a table (best for before/after screenshots) or the <details> disclosure -->

| Before | After |
| - | - |
| <!--insert "before" image here--> | <!--insert "after" image here--> |
| <!--insert "before" image here--> | <!--insert "after" image here--> |
| <!--insert "before" image here--> | <!--insert "after" image here--> |

<details>
<summary>Demo</summary>
<!-- Shorthand image template: <img height=400 src="<URL>" /> -->
</details>

## :pencil: Checklist
- [ ] I filled in the ticket numbers and a description of my work
- [ ] I updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] I ensured unit tests pass and wrote tests for new code
- [ ] If working on UI, I checked and implemented accessibility (Dynamic Text and VoiceOver)
- [ ] If adding telemetry, I read the [data stewardship requirements](https://github.com/mozilla-mobile/firefox-ios/wiki/Adding-Glean-Telemetry-Events) and will request a data review
- [ ] If needed, I updated documentation and added comments to complex code
- [ ] If needed, I added a backport comment (example `@Mergifyio backport release/v120`)

FilippoZazzeroni commented on firefox-ios/Client/Coordinators/LaunchView/ModernLaunchTransitionAnimator.swift:
```diff
@@ -0,0 +1,66 @@
+
+        fadeOutLaunchScreenLoader(fromViewController)
+
+        UIView.animate(
```
I don't think this is the optimal way, we should add a callback to the fade out method, and complete the transition once the fadeout and alpha animations completes

razvanlitianu replied:
Thanks for the feedback! You're absolutely right about the synchronization issue. I've implemented the callback approach you suggested.

FilippoZazzeroni commented on firefox-ios/Client/Coordinators/LaunchView/ModernLaunchScreenViewController.swift:
```diff
@@ -0,0 +1,160 @@
+        super.viewDidLoad()
+        view.backgroundColor = .systemBackground
+
+        Task {
```
can you explain why do we need this ? I don't understand why we need to delay, also how do we make sure the viewWillAppear and viewDidAppear are synced with the Task ? are they independent ?

razvanlitianu commented:
PR updated with unit tests.

FilippoZazzeroni reviewed (commented):
Thanks @razvanlitianu for adding all this tests 🙏 . I added some comments on the implementation, also tagging @cyndichin if she could have an extra eye on the unit test side.

cyndichin commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Coordinators/LaunchView/ModernLaunchScreenViewControllerTests.swift:
```diff
@@ -0,0 +1,246 @@
+    }
+
+    override func tearDown() {
+        super.tearDown()
```
+1 see wiki guidelines here: https://github.com/mozilla-mobile/firefox-ios/wiki/Unit-Tests-Guidelines#1-setupteardown

cyndichin commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Coordinators/LaunchView/ModernLaunchScreenViewControllerTests.swift:
```diff
@@ -0,0 +1,246 @@
+
+    override func tearDown() {
+        super.tearDown()
+        AppContainer.shared.reset()
```
+1

cyndichin commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Coordinators/LaunchView/ModernLaunchScreenViewControllerTests.swift:
```diff
@@ -0,0 +1,246 @@
+        // Simulate loading completion
+        subject.finishedLoadingLaunchOrder()
+
+        let expectation = expectation(description: "Load next launch type called")
```
A couple of these tests seems to aim to be more like integration tests as opposed to unit tests. If there's difficulty in testing unit pieces of work then maybe we can rethink how production code is written too. Using variables like `isLoading \ shouldLoadNextLaunchType` also makes testing a bit awkward. Maybe adding it the view model can be better and we can have a single interface to check without exposing these two booleans.  Therefore, the view controller tests can confirm the call to the method as you done with the other mocks. Let me know if you have any questions!

Also ultimately we would want to call a single method `subject.finishedLoadingLaunchOrder` if we can and test specific logic to that. 

cc: @yoanarios

cyndichin commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Coordinators/LaunchView/ModernLaunchScreenViewControllerTests.swift:
```diff
@@ -0,0 +1,246 @@
+        XCTAssertEqual(coordinatorDelegate.launchWithTypeCalled, 0)
+    }
+
+    func testMemoryManagement_viewControllerIsReleased() {
```
looking at this first glance, I would think `trackForMemoryLeak` would also capture it, so curious to see if there's a specific reason we added this? 🤔

cyndichin commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Coordinators/LaunchView/ModernLaunchScreenViewControllerTests.swift:
```diff
@@ -0,0 +1,246 @@
+        let subject = createModernSubject()
+
+        // Access the view to trigger viewDidLoad naturally
+        _ = subject.view
```
Why not  use `subject.viewDidLoad` since that's where the logic that we are testing. Also, we haven't proposed this officially to the team yet. But we would like to maintain some consistent naming convention. 

@yoanarios and I discussed following something along the lines, so it would be nice to rename while are here to match. Which the tests here follow for the most part but not exactly.

```
func test_init_setsPrefersStatusBarHiddenToTrue() { ... }
func test_viewDidLoad_setsBackgroundColorToSystemBackground() { ... }
func test_viewDidLoad_triggersViewModelStartLoadingOnce() { ... }
func test_viewWillAppear_whenNotLoading_triggersLoadNextLaunchType() { ... }
fun test_finishedLoadingLauchOrder_triggersLoadNextLaunchType() { ... }
```

This is a good reference: https://qualitycoding.org/unit-test-naming/ in general.

cyndichin reviewed (commented):
thanks @razvanlitianu, I am just reviewing from the unit tests perspective. @FilippoZazzeroni covers a couple of things I would mention, but added some more comments. Let me know if there's anything that needs more clarity!!

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Coordinators/LaunchView/ModernLaunchScreenViewControllerTests.swift:
```diff
@@ -0,0 +1,362 @@
+        viewModel.mockLaunchType = .update(viewModel: viewModel.updateViewModel)
+        let subject = createModernSubject()
+
+        _ = subject.view
```
nit: viewDidLoad

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Coordinators/LaunchView/ModernLaunchScreenViewControllerTests.swift:
```diff
@@ -0,0 +1,362 @@
+        viewModel.mockLaunchType = .survey(manager: viewModel.surveySurfaceManager)
+        let subject = createModernSubject()
+
+        _ = subject.view
```
same here

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Coordinators/LaunchView/ModernLaunchScreenViewControllerTests.swift:
```diff
@@ -0,0 +1,362 @@
+        viewModel.mockLaunchType = nil
+        let subject = createModernSubject()
+
+        _ = subject.view
```
same

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Coordinators/LaunchView/ModernLaunchScreenViewControllerTests.swift:
```diff
@@ -0,0 +1,362 @@
+        viewModel.mockLaunchType = .defaultBrowser
+        let subject = createModernSubject()
+
+        _ = subject.view
```
same

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Coordinators/LaunchView/ModernLaunchScreenViewControllerTests.swift:
```diff
@@ -0,0 +1,362 @@
+        XCTAssertTrue(coordinatorDelegate.verifyLaunchWithCallCount(1))
+        XCTAssertTrue(coordinatorDelegate.verifyLaunchBrowserCallCount(1))
+
+        viewModel.reset()
```
not sure about this. @cyndichin should we follow the `arrange-act-assert` pattern right ?
As I interpreted it @razvanlitianu after we fire the tested method, we should assert and then do nothing after it. If we have the needs to add additional call after the final asserts then we need to split the test into multiple once. This helps splitting up the tests and keep the test addressing one test case and not more at the time. let me know if this makes sense to you, cc @razvanlitianu @cyndichin

razvanlitianu replied:
I agree. I removed it entirely because the view model is instantiated every time the setup is executed.

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Mocks/MockLaunchFinishedLoadingDelegate.swift:
```diff
@@ -6,21 +6,90 @@ import Foundation
+        return finishedLoadingLaunchOrderCalled == expectedCount
+    }
+
+    var allLaunchTypes: [LaunchType] {
```
Nit: All the computed properties should go in the property section

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Mocks/MockLaunchScreenManager.swift:
```diff
@@ -32,10 +39,48 @@ class MockLaunchScreenViewModel: LaunchScreenViewModel {
+        return startLoadingCallHistory.last
+    }
+
+    var loadNextLaunchTypeCallCount: Int {
```
Nit: same should go in the properties section

FilippoZazzeroni reviewed (commented):
Thanks @razvanlitianu for addressing all the comments, I just added a bunch of comments, most of them are just nits. Let me know for anything

razvanlitianu commented:
@FilippoZazzeroni @cyndichin

Thanks for taking the time to review the PR. I’ve made several updates and wanted to summarize the changes:
Removed unnecessary self references.
Removed some helper methods since the same checks can be done with assertions directly.
I ran into issues with initializing the view controller and triggering viewDidLoad. Calling it manually caused the view model methods to run twice. The reliable approach I found was to access subject.view, which internally triggers viewDidLoad. In the end, I removed those manual calls and now only call either the view model methods or the view controller methods directly.
Removed the reset methods after realizing the view model is re-instantiated for each test anyway.

razvanlitianu commented on firefox-ios/Client/Coordinators/LaunchView/ModernLaunchScreenViewController.swift:
```diff
@@ -0,0 +1,148 @@
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        view.backgroundColor = .systemBackground
```
Yes, great find. It was carried over from LaunchScreenViewController. Since I don't use it, I will just remove it.

cyndichin commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Coordinators/LaunchView/ModernLaunchScreenViewControllerTests.swift:
```diff
@@ -0,0 +1,362 @@
+        XCTAssertEqual(coordinatorDelegate.allLaunchTypes.count, 4)
+    }
+
+    private func createModernSubject(file: StaticString = #filePath,
```
+1

FilippoZazzeroni reviewed (commented):
@cyndichin sorry I haven't loop you in on that, we had a sync me and @razvanlitianu on the controller and agreed to have a follow up ticket to refactor the `ModernLaunchScreenViewController`. @razvanlitianu can u please address Cyndi's comment on the tests side and add in the `ModernLaunchScreenViewController` the follow up ticket for the refactor.

razvanlitianu commented:
Updated the code and created https://mozilla-hub.atlassian.net/browse/FXIOS-13434

FilippoZazzeroni reviewed (approved):
LGTM @razvanlitianu thanks for addressing all the comments 🙏
